### ISSUE TITLE
symput does not allow me to put one underscore for a macro var

### ISSUE THREAD

**Issue by sulnm019** [sassoftware/saspy#269]
Hello, I've tried many different ways just to be able to put **10_2019** as a string value for symput.
However, every time I get the macro var, the underscore ends up being dropped and it looks like **102019**.
I've tried adding a double underscore and a few other alternatives, but for some reason it won't work.
I need it to be formatted like 10_2019 to be able to access data.

Below are examples of what I've tried.
![image](https://user-images.githubusercontent.com/55457502/67024196-0178c080-f0d2-11e9-91c4-de36c8aebf25.png)

![image](https://user-images.githubusercontent.com/55457502/67024319-3422b900-f0d2-11e9-96d5-e9714cdab69e.png)

**Comment by tomweber-sas** [comment]
Hey, I can help with that. I tried this simply with the description you said, not the other more complicated cases. It worked as expected. One thing that is helpful is to look at the log to see what's happening. I recently added a new method that's helpful here, lastlog() (as opposed to saslog()), which shows the log from the last subit that was executed. symput/symget will simply issue a submit, so after either of those, print(sas.lastlog()) will show you just what they did.

Here's what I ran and the string sees to be valid all the way through. Can you look at the log for your cases to see if there's any clues in there?

```
>>> sas.symput('t1', '10_1920')
>>> print(sas.lastlog())
12                                                         The SAS System                           10:49 Thursday, October 17, 2019

137        ods listing close;ods html5 (id=saspy_internal) file=_tomods1 options(bitmap_mode='inline') device=svg style=HTMLBlue;
137      ! ods graphics on / outputfmt=png;
NOTE: Writing HTML5(SASPY_INTERNAL) Body file: _TOMODS1
138
139        %let t1=%NRBQUOTE(10_1920);
1                                                          The SAS System                           12:14 Thursday, October 17, 2019

NOTE: html5 Memory Usage (274 events)
      Total: 2.79MB     High: 1283.89KB     Max: 32807B     Allocs: 16839   Frees: 16839
140
141
142        ods html5 (id=saspy_internal) close;ods listing;
143
>>> print(sas.submit("%put &t1;")['LOG'])
13                                                         The SAS System                           10:49 Thursday, October 17, 2019

146        ods listing close;ods html5 (id=saspy_internal) file=_tomods1 options(bitmap_mode='inline') device=svg style=HTMLBlue;
146      ! ods graphics on / outputfmt=png;
NOTE: Writing HTML5(SASPY_INTERNAL) Body file: _TOMODS1
147
148        %put &t1;
10_1920
1                                                          The SAS System                           12:15 Thursday, October 17, 2019

NOTE: html5 Memory Usage (274 events)
      Total: 2.79MB     High: 1283.89KB     Max: 32807B     Allocs: 16839   Frees: 16839
149
150        ods html5 (id=saspy_internal) close;ods listing;
151
>>> t1 = sas.symget('t1')
>>> t1
'10_1920'
>>>
```

**Comment by sulnm019** [comment]
okay so i tried that,
![image](https://user-images.githubusercontent.com/55457502/67029488-68e73e00-f0db-11e9-8751-f62fa658f885.png)
![image](https://user-images.githubusercontent.com/55457502/67029524-78668700-f0db-11e9-9a23-0331ad09c020.png)
![image](https://user-images.githubusercontent.com/55457502/67029554-887e6680-f0db-11e9-9b54-f8ce05742aaa.png)


and i seem to still be getting the wrong format

**Comment by tomweber-sas** [comment]
Hmmm, I tried that using the IOM access method, in line mode. As you see it worked correctly. I just tried the same thing in a jupyter notebook, also using IOM, and everything was the same except that last line where t1 no longer showed the underscore. I'll need to debug this in jupyter to see what's going on.
```
sas.symput('t1', '10_1920')
print(sas.lastlog())

3                                                          The SAS System                           12:36 Thursday, October 17, 2019

20         ods listing close;ods html5 (id=saspy_internal) file=_tomods1 options(bitmap_mode='inline') device=svg style=HTMLBlue;
20       ! ods graphics on / outputfmt=png;
NOTE: Writing HTML5(SASPY_INTERNAL) Body file: _TOMODS1
21         
22         %let t1=%NRBQUOTE(10_1920);
23         
24         
25         ods html5 (id=saspy_internal) close;ods listing;
26         


print(sas.submit("%put &t1;")['LOG'])

4                                                          The SAS System                           12:36 Thursday, October 17, 2019

29         ods listing close;ods html5 (id=saspy_internal) file=_tomods1 options(bitmap_mode='inline') device=svg style=HTMLBlue;
29       ! ods graphics on / outputfmt=png;
NOTE: Writing HTML5(SASPY_INTERNAL) Body file: _TOMODS1
30         
31         %put &t1;
10_1920
32         
33         ods html5 (id=saspy_internal) close;ods listing;
34         

t1 = sas.symget('t1')
t1

101920
```

And, the following is also the same for both: so there must be something strange going on here.
```
print(sas.lastlog())


5                                                          The SAS System                           12:36 Thursday, October 17, 2019

37         ods listing close;ods html5 (id=saspy_internal) file=_tomods1 options(bitmap_mode='inline') device=svg style=HTMLBlue;
37       ! ods graphics on / outputfmt=png;
NOTE: Writing HTML5(SASPY_INTERNAL) Body file: _TOMODS1
38         
39         %put t1=&t1 tom=;
t1=10_1920 tom=
40         
41         
42         ods html5 (id=saspy_internal) close;ods listing;
43         
​```

**Comment by tomweber-sas** [comment]
well, ok, this seems to be a discrepancy in python itself (my guess at the moment). meaning behavior has changed between versions. Here's why.

for symget, I get a string, as I'm parsing the log with the value of the macro variable written to the log; so, by definition I have a string. But, maybe that string it '42' and it's really a number that you're using. So, I want to return the correct python type so if it's really a number you can math on it or whatever :)

Here my code:

```
    def symget(self, name):
        """
        :param name:  name of the macro varable to set:

            - name    is a character

        """
        ll = self.submit("%put " + name + "=&" + name + " tom=;\n")

        l2 = ll['LOG'].rpartition(name + "=")[2].partition(" tom=")
        try:
            var = int(l2[0])
        except:
            try:
                var = float(l2[0])
            except:
                var = l2[0]

        return var
```

When I run this on my linux system (different python install than on windows) and where I really do my development, I get the 'right' behavior:
```
>>> t1
'10_1920'
>>>
>>> try:
...     var = int(t1)
... except:
...     try:
...         var = float(t1)
...     except:
...         var = t1
...
>>>
>>>
>>> var
'10_1920'
>>> int('10_1920')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: invalid literal for int() with base 10: '10_1920'
>>>
```

But, when I run on windows, anaconda something, different version at least, then on my linux, and it's not jupyter because I can run jupyter on either python and it's still different results depending upon the python, I get the different behavior (which seems incorrect to me, as '10_1982' isn't an int):

```
t1 = '10_1920'

try:                                 
    var = int(t1)                 
except:                              
    try:                             
        var = float(t1)           
    except:                          
        var = t1

var
101920

type(var)
int

int('10_1920')
101920
```

So, this is troublesome, as a particular version of python considers a sting with underscores to be an int, while another version doesn't. I can't fix that.

But, what I can do, is add a 2nd parameter to my symget method, which lets you specify a type for the return value, say, [str, float, int] and if you specify that, I'll convert to it outright, else do what I'm doing now, which should (has) generally work(ed). 

What do you think about that as a solution to this? It's what I have control over, and can do.

for this you would do
sas.symput('date', '10_1920')
sas.symget('date', str())    # instead of sas.symget('date')

The type would be a python type (int() or 3,  str() or '' or 'a', float() or 1.0  ...:  
>>> type(int())
<class 'int'>
>>> type(str())
<class 'str'>
>>> type(float())
<class 'float'>
>>> type(3)
<class 'int'>
>>> type(3.0)
<class 'float'>
>>> type('')
<class 'str'>



Tom

**Comment by sulnm019** [comment]
I like that solution. I feel like it would be useful for other windows users, using a different IDE than Jupyter.

**Comment by tomweber-sas** [comment]
yes, and I don't think it's the ide or the OS, I think it has to be pyyton itself. Here's in terminal mode from my anaconda on windows: 

I'll enhance sysmget, as I also think it could be useful in other situation to explicitly say what type you want it back as. May you want '42' as a string and not an int ...

This won't take long, and I can push it to master for you so you can try it out. 

```
C:\Users\sastpw>python
Python 3.6.0 |Anaconda 4.3.0 (64-bit)| (default, Dec 23 2016, 11:57:41) [MSC v.1900 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>>
>>>
>>> t1 = '10_1920'
>>>
>>> try:
...     var = int(t1)
... except:
...     try:
...         var = float(t1)
...     except:
...         var = t1
...
>>>
>>> var
101920
>>> type(var)
<class 'int'>
>>>
```

**Comment by tomweber-sas** [comment]
Ok, I have this coded up. This is pushed to master. Give it a try and see that it works like you would like. Let me know if you have any other suggestions for it.

I support the same 3 types I return already, int float str. If you don't specify one, it's the same as before, of course (no breaking changes here). If you specify one, I use it, and if it's not able to convert to that, you get the error; I don't think I should try to change it to something else when you explicitly specify one;  specify one that is valid and it works :), else you should see the error.

If you need some other type, you can probably convert it after this. If you has a string that had a bunch of names or something, you could convert that to a list yourself after w/ one python statement. From SAS a numeric or a string is about all you would really get, and int is just a helpful version of numeric when it's really not a floating point. 

Here's a list post process example:

```
>>> sas.symput('t1', 'bob fred sally someoneelse')
>>> t1 = sas.symget('t1')
>>> type(t1)
<class 'str'>
>>> t1
'bob fred sally someoneelse'
>>> t1.split()
['bob', 'fred', 'sally', 'someoneelse']
>>> type(t1.split())
<class 'list'>
>>>
```

Here's a little more output:

```
>>> sas.symput('t1', 10.1920)
>>> t1 = sas.symget('t1'); t1; type(t1)
10.192
<class 'float'>
>>>
>>> t1 = sas.symget('t1', 1); t1; type(t1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/opt/tom/github/saspy/saspy/sasbase.py", line 1344, in symget
    var = int(l2[0])
ValueError: invalid literal for int() with base 10: '10.192'
>>>
>>> t1 = sas.symget('t1', 1.0); t1; type(t1)
10.192
<class 'float'>
>>>
>>> t1 = sas.symget('t1', '1.0'); t1; type(t1)
'10.192'
<class 'str'>
>>>
>>> t1 = sas.symget('t1', []); t1; type(t1)
invalid type specified. supported are [int, float, str], will return default type
10.192
<class 'float'>
>>>
>>> t1 = sas.symget('t1', int()); t1; type(t1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/opt/tom/github/saspy/saspy/sasbase.py", line 1344, in symget
    var = int(l2[0])
ValueError: invalid literal for int() with base 10: '10.192'
>>>
>>> t1 = sas.symget('t1', float()); t1; type(t1)
10.192
<class 'float'>
>>>
>>> t1 = sas.symget('t1', str()); t1; type(t1)
'10.192'
<class 'str'>
>>>
>>> t1 = sas.symget('t1', list); t1; type(t1)
invalid type specified. supported are [int, float, str], will return default type
10.192
<class 'float'>
>>>
```

**Comment by tomweber-sas** [comment]
Hey, I finally had a chance to investigate this and it appears that this is a change in python as of version 3.6. 


From the python doc for int()
```
class int(x=0)
class int(x, base=10)
...
[same definition as in previous versions]
...
Changed in version 3.6: Grouping digits with underscores as in code literals is allowed.

```

Looking at what **Grouping digits with underscores as in code literals** is defined to mean I hit lots of info on this. Here's a couple excerpts from https://www.python.org/dev/peps/pep-0515/

```
PEP 515 -- Underscores in Numeric Literals

# grouping decimal numbers by thousands
amount = 10_000_000.0

# grouping hexadecimal addresses by words
addr = 0xCAFE_F00D

# grouping bits into nibbles in a binary literal
flags = 0b_0011_1111_0100_1110

# same, for string conversions
flags = int('0b_1111_0000', 2)
```

So, it appears this is a valid int, so maybe just stay away from this syntax in the future for variable names. And, it seems clear that allowing you to explicitly specify your intended output type from symget() is good thing to have.

If this is working for you, we can close this. If there's anything else you need, just let me know.

Thanks!
Tom

**Comment by tomweber-sas** [comment]
This is now available in the current production version - 3.1.7. I'll go ahead an d close this. Let me know if you have any other issues!

Thanks,
Tom
